realm/realm-library/src/main/java/io/realm/RealmList.java

```diff
@@ -1288,15 +1292,1274 @@ public void add(@Nullable E e) {
+        if (value == null) {
+            // null is always valid (but schema may reject null on insertion).
+            return;
+        }
+        if (!(value instanceof Number)) {
+            throw new IllegalArgumentException(
+                    String.format(Locale.ENGLISH, INVALID_OBJECT_TYPE_MESSAGE,
+                            "java.lang.Number",
+                            value.getClass().getName()));
+        }
+    }
+
+    @Override
+    public void add(@Nullable Object value) {
+        checkValidValue(value);
+        if (value == null) {
+            osList.addNull();
+        } else {
+            osList.addFloat(((Number) value).floatValue());
+        }
+    }
+
+    @Override
+    public void insert(int index, @Nullable Object value) {
+        checkValidValue(value);
+        checkInsertIndex(index);
+
+        if (value == null) {
+            osList.insertNull(index);
+        } else {
+            osList.insertFloat(index, ((Number) value).floatValue());
+        }
+    }
+
+    @Nullable
+    @Override
+    protected Float set(int index, @Nullable Object value) {
+        checkValidValue(value);
+        checkElementExists(index);
+
+        final Float oldValue = get(index);
+        if (value == null) {
+            osList.setNull(index);
+        } else {
+            osList.setFloat(index, ((Number) value).floatValue());
+        }
+        return oldValue;
+    }
+}
+
+final class DateListOperator extends ManagedListOperator<Date> {
+
+    DateListOperator(BaseRealm realm, OsList osList, Class<Date> clazz) {
+        super(realm, osList, clazz);
+    }
+
+    @Override
+    boolean forRealmModel() {
+        return false;
+    }
+
+    @Nullable
+    @Override
+    Date get(int index) {
+        return (Date) osList.getValue(index);
+    }
+
+    private void checkValidValue(@Nullable Object value) {
+        if (value == null) {
+            // null is always valid (but schema may reject null on insertion).
+            return;
+        }
+        if (!(value instanceof Date)) {
+            throw new IllegalArgumentException(
+                    String.format(Locale.ENGLISH, INVALID_OBJECT_TYPE_MESSAGE,
+                            "java.util.Date",
+                            value.getClass().getName()));
+        }
+    }
+
+    @Override
+    public void add(@Nullable Object value) {
+        checkValidValue(value);
+        if (value == null) {
+            osList.addNull();
+        } else {
+            osList.addDate((Date) value);
+        }
+    }
+
+    @Override
+    public void insert(int index, @Nullable Object value) {
+        checkValidValue(value);
+        checkInsertIndex(index);
+
+        if (value == null) {
+            osList.insertNull(index);
+        } else {
+            osList.insertDate(index, (Date) value);
+        }
+    }
+
+    @Nullable
+    @Override
+    protected Date set(int index, @Nullable Object value) {
+        checkValidValue(value);
```

Comment:
the type needs to be checked any way. So if
```java
abstract Class checkValidValue(value);

T set(int index, @Nullable Object value) {
    Class clazz = checkValidValue(value);
    if (value == null) {
        osList.setNull(index, value);
        return;
    }

    switch(clazz) {
        case Date.class:
            osList.setDate();
            return;
        case ...:
    }
}
```

Then the child class only need to implement `checkValidValue()`.

Reply:
refactored